felix/labelindex/named_port_index.go

```diff
@@ -568,48 +646,56 @@ func (idx *SelectorAndNamedPortIndex) UpdateEndpointOrSet(
 			idx.discardParentIfEmpty(parent.id)
 		}
 	}
+
+	gaugeNumEndpoints.Set(float64(idx.endpointKVIdx.Len()))
 }
 
-func (idx *SelectorAndNamedPortIndex) scanEndpointAgainstAllIPSets(
+func (idx *SelectorAndNamedPortIndex) scanEndpointAgainstIPSets(
 	epData *endpointData,
 	oldIPSetContributions map[string][]IPSetMember,
 ) {
-	for ipSetID, ipSetData := range idx.ipSetDataByID {
+	// Remove any previous match from the endpoint's cache.  We'll re-add it
```

Comment:
This method needed rework because we no longer scan all selectors. If the labels have changes there's no guarantee that we'd scan all the previously matching selectors so we now use `oldIPSetContributions` to do the decrefs at the end.